service/qos/raft_service_level_distributed_data_accessor.cc

```diff
@@ -0,0 +1,101 @@
+                    days_counter{0},
+                    nanoseconds_counter{std::chrono::duration_cast<std::chrono::nanoseconds>(d).count()}));
+        },
+    }, tv);
+}
+
+static future<std::vector<mutation>> get_set_mutations(cql3::query_processor& qp, api::timestamp_type timestamp, sstring service_level_name, qos::service_level_options slo) {
+    static sstring insert_query = format("INSERT INTO {}.{} (service_level, timeout, workload_type) VALUES (?, ?, ?);", db::system_keyspace::NAME, db::system_keyspace::SERVICE_LEVELS_V2);
+    data_value workload = slo.workload == qos::service_level_options::workload_type::unspecified
+            ? data_value::make_null(utf8_type)
+            : data_value(qos::service_level_options::to_string(slo.workload));
+    
+    co_return co_await qp.get_mutations_internal(insert_query, qos_query_state(), timestamp, {service_level_name, timeout_to_data_value(slo.timeout), workload});
+}
+
+static future<std::vector<mutation>> get_delete_mutations(cql3::query_processor& qp, api::timestamp_type timestamp, sstring service_level_name) {
+    static sstring delete_query = format("DELETE FROM {}.{} WHERE service_level= ?;", db::system_keyspace::NAME, db::system_keyspace::SERVICE_LEVELS_V2);
+    co_return co_await qp.get_mutations_internal(delete_query, qos_query_state(), timestamp, {service_level_name});
+}
+
+raft_service_level_distributed_data_accessor::raft_service_level_distributed_data_accessor(cql3::query_processor& qp, service::raft_group0_client& group0_client)
+    : _qp(qp)
+    , _group0_client(group0_client) {}
+
+future<qos::service_levels_info> raft_service_level_distributed_data_accessor::get_service_levels() const {
+    return qos::get_service_levels(_qp, db::system_keyspace::NAME, db::system_keyspace::SERVICE_LEVELS_V2, db::consistency_level::LOCAL_ONE);
+}
+
+future<qos::service_levels_info> raft_service_level_distributed_data_accessor::get_service_level(sstring service_level_name) const {
+    return qos::get_service_level(_qp, db::system_keyspace::NAME, db::system_keyspace::SERVICE_LEVELS_V2, std::move(service_level_name), db::consistency_level::LOCAL_ONE);
+}
+
+future<> raft_service_level_distributed_data_accessor::do_raft_command(std::optional<service::group0_guard> guard, std::function<future<std::vector<mutation>>(api::timestamp_type)> mutation_function, std::string_view description, std::vector<mutation> additional_mutations) const {
+    if (this_shard_id() != 0) {
+        on_internal_error(logger, "raft_service_level_distributed_data_accessor: must be executed on shard 0");
+    }
+
+    if (!guard) {
+        on_internal_error(logger, "raft_service_level_distributed_data_accessor: guard must be present");
+    }
+    
+    //NOTE: abort_source here is temporary. It's fixed and passed as an argument in a later commit
+    abort_source as;
+    auto timestamp = guard->write_timestamp();
+
+    auto muts = co_await mutation_function(timestamp);
```

Comment:
if you don't need to do retries I think you can simplify and use vector of mutations instead of callback.

Reply:
You're right, it's a leftover from version where I obtained group0 guard in raft DDA. Fixed